Add `fetchpriority="high"` when next/image has `priority` prop

sanjaiyan-dev opened the pull request:
Eagerly load images on viewport 🖼️ 

> When user scroll to bottom of the website so swiftly (which have lot of images) currently `Next.js` marks even other image also needed to be fetched which might cause delay of loading of current image which is located in user's current position.
> So, we can add [priority hint](https://web.dev/priority-hints) to the particular image element in the viewport to load in high priority.

Extremely sorry my explanation might not be good, hope you got what I am trying to say 😆 

_Sorry if I made any mistakes :(_

SukkaW reviewed (changes requested):
According to the WICG Priority Hints draft:

> With in-viewport images, a browser may automatically detect if the image is important and boost priority if it gets to layout early enough.

As the draft described, the browser could have detected the in-view image and automatically boost the fetch priority, it is probably unnecessary to dynamically set the `fetchpriority`.

Instead of dynamically setting `fetchpriority` to `high` when intersected and removing the `fetchpriority` when out of the viewport, personally I'd prefer setting a static `fetchpriority`.

`next/image` component already accepts the `priority={boolean}` prop. Would you like to change the implementation so that the `fetchpriority="high"` is set when `priority={true}` is specified?

sanjaiyan-dev commented:
> According to the WICG Priority Hints draft:
> 
> > With in-viewport images, a browser may automatically detect if the image is important and boost priority if it gets to layout early enough.
> 
> As the draft described, the browser could have detected the in-view image and automatically boost the fetch priority, it is probably unnecessary to dynamically set the `fetchpriority`.
> 
> Instead of dynamically setting `fetchpriority` to `high` when intersected and removing the `fetchpriority` when out of the viewport, personally I'd prefer setting a static `fetchpriority`.
> 
> `next/image` component already accepts the `priority={boolean}` prop. Would you like to change the implementation so that the `fetchpriority="high"` is set when `priority={true}` is specified?

Understood :)

styfle reviewed (changes requested):
> Would you like to change the implementation so that the fetchpriority="high" is set when priority={true} is specified

Let's not do that without measuring the impact on performance. The last I checked, the `fetchpriority` prop is not as good as the current `priority` preloading prop.

sanjaiyan-dev commented:
https://hacktoberfest.com/participation/#pr-mr-details

sanjaiyan-dev commented on packages/next/client/future/image.tsx:
```diff
@@ -349,6 +350,7 @@ const ImageElement = ({
         className={className}
         // @ts-ignore - TODO: upgrade to `@types/react@17`
         loading={loading}
+        fetchPriority={priority ? "high" : null}
```
> IMHO `undefined` should be used as the default value.
> 
> Also, let's add a comment with the `web.dev` link, which explains why `fetchPriority` is set even when `<link rel=preload>` is already used.
> 
> I will also try to deploy a demo, so that we could perform some web vitals benchmarks to see if adding `fetchPriority` really helps.

Ok :)

sanjaiyan-dev commented:
> > Would you like to change the implementation so that the fetchpriority="high" is set when priority={true} is specified
> 
> Let's not do that without measuring the impact on performance. The last I checked, the `fetchpriority` prop is not as good as the current `priority` preloading prop.

Additional ref-: https://angular.io/guide/image-directive#marking-images-as-priority

styfle reviewed (commented):
Does the `fetchPriority` prop need to be added to the preload `<link>` in the head?

sanjaiyan-dev commented:
> Does the `fetchPriority` prop need to be added to the preload `<link>` in the head?

@styfle I think no need according to implementation of angular.

styfle commented:
Looks like CI is failing due to missing TS types

Also, needs updated tests to ensure the prop is working correctly

styfle commented on packages/next/client/future/image.tsx:
```diff
@@ -351,6 +352,8 @@ const ImageElement = ({
         // @ts-ignore - TODO: upgrade to `@types/react@17`
         loading={loading}
+        // https://web.dev/priority-hints/#increase-the-priority-of-the-lcp-image
+        fetchPriority={priority ? "high" : undefined}
```
undefined should probably be changed to `rest.fetchPriority` so we don't break anyone who is currently setting this prop to a custom value like `"low"` for example.

sanjaiyan-dev replied:
> undefined should probably be changed to `rest.fetchPriority` so we don't break anyone who is currently setting this prop to a custom value like `"low"` for example.

Done :)

balazsorban44 commented:
@sanjaiyan-dev is there an intention to push this forward? Currently, there are merge conflicts and there are unaddressed comments here: https://github.com/vercel/next.js/pull/40887#issuecomment-1276848121

sanjaiyan-dev commented:
> @sanjaiyan-dev is there an intention to push this forward? Currently, there are merge conflicts and there are unaddressed comments here: https://github.com/vercel/next.js/pull/40887#issuecomment-1276848121

Yh, I will fix it ASAP 🤝🙌

styfle commented:
How does this affect performance of stylesheets? Do we need to configure fetchPriority on those too?

sanjaiyan-dev commented:
> How does this affect performance of stylesheets? Do we need to configure fetchPriority on those too?

Extremely sorry I am not sure about it but I created PR related loading of first party scripts and optimized way of preloading.

https://github.com/vercel/next.js/pull/41084

https://github.com/vercel/next.js/pull/41423

sanjaiyan-dev commented:
Let's Start from scratch :)

sanjaiyan-dev commented:
> @sanjaiyan-dev is there an intention to push this forward? Currently, there are merge conflicts and there are unaddressed comments here: [#40887 (comment)](https://github.com/vercel/next.js/pull/40887#issuecomment-1276848121)

I have created refined PR 🤝 

https://github.com/vercel/next.js/pull/43096
